### ISSUE TITLE
GVAE node features and padding 

### ISSUE THREAD

**Issue by ahariri13** [rusty1s/pytorch_geometric#1232]
I am facing a set of doubts when working on a Graph VAE to learn on batches of graphs and I was hoping I could get some clarifications on some points: 

1- In my problem I care mainly about generating node features for point clouds rather than their adjacency matrix. That being said, should I include the adjacency matrix in the loss function or train to reconstruct solely the feature matrix (while keeping the Adj Matrix fixed).

2-It was mentioned in issue #913 that we do not do zero-nodes padding but rather stack A's diagonally. Does the same apply to GVAEs ?  In my case every graph in the batch has a different number of nodes, what I did so far is took the max number of nodes and zero-padded to it then only accounted for non-zero nodes using a custom loss function I wrote. 

Any suggestion to prevent padding would be appreciated because despite my decreasing loss in training, my generated values have a variance that is too small, I expected a larger range of values, so I think it could partly be due to padding. 

Thanks a lot in advance ! :)

**Comment by rusty1s** [comment]
IMO, you do not really need padding when you only want to generate node features (vertically stacked). You only need padding in case you reconstruct adjacency matrices, as you cannot reconstruct an adjacency matrix that is diagonally stacked (since this would require predicting the existence of an edge between nodes of different example).

Concerning additional loss information, I doubt that this will bring any gain when you solely want to generate node features.

**Comment by ahariri13** [comment]
Thanks a lot @rusty1s for your feedback. I eventually did it without padding though I have one final concern which is mainly theoretical. I wrote the following code to reconstruct the graph features, and it is giving reasonably good reconstructions. The only thing I have in mind is that my graph nodes (around 500 each)  consist of 3 features only. 

What I did basically is transformed them into a hidden space using linear transformations and applied Gated Graph Sequence, then transformed them again using another linear transform.  To decode them I used a FCN and get my reconstructed features using the KL divergence and reconstruction errors in the loss function. 

In theory, an auto-encoder works mostly by compressing the input into a smaller dimension. But my nodes have only 3 features (unlike the citation network example which have around 2000 if I remember well) so I cannot go smaller than that reasonably. Is it wrong what I did by working in higher dimensions ? 

Thanks a lot in advance!

```
class GCN_Message(nn.Module):
    def __init__(self,in_channels, out_channels1, out_channels2,deco2,deco3,deco4,num_layers, dropout,batch_size):
        super(GCN_Message, self).__init__()   
        self.lin1=nn.Linear(in_channels,150)#out_channels1)
        self.lin2=nn.Linear(150,out_channels1)

        
        self.dec1=nn.Linear(64,deco2)
        self.dec2=nn.Linear(deco2,deco3)
        self.dec3=nn.Linear(deco3,deco4)
        self.dec4=nn.Linear(deco4,3)


        self.out_channels2=out_channels2
        self.bn1 = nn.BatchNorm1d(num_features=deco2)
        self.bn2=nn.BatchNorm1d(num_features=deco3)
        """
        Gated GNN
        """
        self.ggn=tnn.GatedGraphConv(out_channels1,num_layers,aggr='mean')
        self.tr1=nn.Linear(out_channels1,out_channels2)
        self.tr2=nn.Linear(out_channels2,64)
        
        

    def encode(self,x,adj):
        trans1=self.lin1(x)
        trans1=F.tanh(trans1)
        hidden1=self.ggn(trans1,adj)
        hidden2=self.tr1(hidden1)
        return self.tr2(hidden2),self.tr2(hidden2) 

    def reparametrize(self, mu, logvar):
      if self.training:
          return mu + torch.randn_like(logvar) * torch.exp(logvar)
      else:
          return mu  

    
    def decode(self,z):
        out1=self.dec1(z)
        out1=F.leaky_relu(out1)
        out1=F.dropout(out1)
        out1=self.bn1(out1)
        out2=self.dec2(out1)
        out2=F.leaky_relu(out2)
        out2=F.dropout(out2)
        out2=self.bn2(out2)
        out2=self.dec3(out2)
        out2=F.tanh(out2)
        out2=self.dec4(out2)

        return out2
        
    
    def forward(self,x,adj):
        mu, logvar = self.encode(x,adj)   
        z = self.reparametrize(mu, logvar) 
        z2=self.decode(z)
        return z2[:,[0,1]],z2[:,2], mu, logvar  
```

**Comment by ahariri13** [comment]
Hi @rusty1s thanks a lot for your feedback. I eventually did it without padding though I have one final concern which is mainly theoretical. I wrote the following code to reconstruct the graph features, and it is giving reasonably good reconstructions. The only thing I have in mind is that my graph nodes (around 500 each)  consist of 3 features only. 

What I did basically is transformed them into a hidden space using linear transformations and applied Gated Graph Sequence, then transformed them again using another linear transform.  To decode them I used a FCN and get my reconstructed features using the KL divergence and reconstruction errors in the loss function. 

In theory, an auto-encoder works mostly by compressing the input into a smaller dimension. But my nodes have only 3 features (unlike the citation network example which have around 2000 if I remember well) so I cannot go smaller than that reasonably. Is it wrong what I did by working in higher dimensions ? 

Thanks a lot in advance!

```
class GCN_Message(nn.Module):
    def __init__(self,in_channels, out_channels1, out_channels2,deco2,deco3,deco4,num_layers, dropout,batch_size):
        super(GCN_Message, self).__init__()   
        self.lin1=nn.Linear(in_channels,150)#out_channels1)
        self.lin2=nn.Linear(150,out_channels1)

        
        self.dec1=nn.Linear(64,deco2)
        self.dec2=nn.Linear(deco2,deco3)
        self.dec3=nn.Linear(deco3,deco4)
        self.dec4=nn.Linear(deco4,3)


        self.out_channels2=out_channels2
        self.bn1 = nn.BatchNorm1d(num_features=deco2)
        self.bn2=nn.BatchNorm1d(num_features=deco3)
        """
        Gated GNN
        """
        self.ggn=tnn.GatedGraphConv(out_channels1,num_layers,aggr='mean')
        self.tr1=nn.Linear(out_channels1,out_channels2)
        self.tr2=nn.Linear(out_channels2,64)
        
        

    def encode(self,x,adj):
        trans1=self.lin1(x)
        trans1=F.tanh(trans1)
        hidden1=self.ggn(trans1,adj)
        hidden2=self.tr1(hidden1)
        return self.tr2(hidden2),self.tr2(hidden2) 

    def reparametrize(self, mu, logvar):
      if self.training:
          return mu + torch.randn_like(logvar) * torch.exp(logvar)
      else:
          return mu  

    
    def decode(self,z):
        out1=self.dec1(z)
        out1=F.leaky_relu(out1)
        out1=F.dropout(out1)
        out1=self.bn1(out1)
        out2=self.dec2(out1)
        out2=F.leaky_relu(out2)
        out2=F.dropout(out2)
        out2=self.bn2(out2)
        out2=self.dec3(out2)
        out2=F.tanh(out2)
        out2=self.dec4(out2)

        return out2
        
    
    def forward(self,x,adj):
        mu, logvar = self.encode(x,adj)   
        z = self.reparametrize(mu, logvar) 
        z2=self.decode(z)
        return z2[:,[0,1]],z2[:,2], mu, logvar  
```

**Comment by rusty1s** [comment]
I agree. Encoding a [N, 3] matrix in a [N, d] matrix is not particularly useful (and trivial if you do not care about graph connectivity). For point cloud compression, it is more reasonable to think about compressing the node dimension. There is also a lot of literature about that, *e.g.*, https://arxiv.org/pdf/1707.02392.pdf., which should be quite straightforward to implement.

**Comment by ahariri13** [comment]
Hey, Sorry for the late reply. Meanwhile, I have tried several modifications of Graph-UNet in order to make a Variational Graph Auto-Encoder (rather than a regular GAE). Yet, I feel the process of retaining  **(xs,edge_indices,edge_weights,perms)** between my encoder and decoder makes the task much slower: My GPU utilization stays relatively low. In the code below, I first linearly transform my graph features, then perform Gated GraphConv followed by SAGPooling many times. I retain the indices of the nodes, the edge indices, weights and perms and use them again in the decoder. I also tried to make sure to free my GPU memory using the commands visible in the decoder (but I don't think they had any effect).  Could I get your opinion on how correct my methodology is / what I could improve to get better reconstructions for my node features ? 

`import torch
import torch.nn.functional as F
from torch_sparse import spspmm
from torch_geometric.nn import TopKPooling, GCNConv,GatedGraphConv, SAGPooling
from torch_geometric.utils import (add_self_loops, sort_edge_index,
                                   remove_self_loops)
from torch_geometric.utils.repeat import repeat



class GraphUNet(torch.nn.Module):
    def __init__(self, in_channels, hidden_channels, out_channels, depth,
                 pool_ratios, sum_res=True, act=F.tanh):
        super(GraphUNet, self).__init__()
        assert depth >= 1
        self.in_channels = in_channels
        self.hidden_channels = hidden_channels
        self.out_channels = out_channels
        self.depth = depth
        self.pool_ratios = repeat(pool_ratios, depth)
        self.act = act
        self.act2=F.leaky_relu
        self.sum_res = sum_res

        channels = hidden_channels
        self.trans=torch.nn.Linear(in_channels,64)
        self.down_convs = torch.nn.ModuleList()
        self.pools = torch.nn.ModuleList()
        self.down_convs.append(GCNConv(64, int(channels), improved=True))
        for i in range(depth):
            if i<depth-1: 
                self.pools.append(SAGPooling(channels, self.pool_ratios[i]).cuda())
                self.down_convs.append(GatedGraphConv(int(channels*2),num_layers=2,aggr='mean').cuda())
                channels=int(channels*2)
            else:
              self.pools.append(SAGPooling(channels, self.pool_ratios[i]).cuda())
              self.down_convs.append(GatedGraphConv(int(channels),num_layers=2,aggr='mean').cuda())
              
        in_channels = channels if sum_res else 2 * channels

        self.up_convs = torch.nn.ModuleList()
        for i in range(depth - 1):
            self.up_convs.append(GCNConv(channels, int(channels/2), improved=True).cuda())
            channels=int(channels/2)
        self.up_convs.append(GCNConv(channels, out_channels, improved=True).cuda())


        self.muLay=torch.nn.Linear(256,512)
        self.sigLay=torch.nn.Linear(256,512)
        
        
        self.dec=torch.nn.Linear(512,256)
        self.drop=torch.nn.Dropout(p=0.2)
        
    def encode(self,x,edge_index,batch): 
        x=self.trans(x)
        if batch is None:
            batch = edge_index.new_zeros(x.size(0))
        edge_weight = x.new_ones(edge_index.size(1))

        x = self.down_convs[0](x, edge_index, edge_weight)
        x = self.act(x)

        xs = [x]
        edge_indices = [edge_index]
        edge_weights = [edge_weight]
        perms = []

        for i in range(1, self.depth + 1): 
            edge_index, edge_weight = self.augment_adj(edge_index, edge_weight,x.size(0))
            
            x, edge_index, edge_weight, batch, perm, _ = self.pools[i - 1](
                x, edge_index, edge_weight, batch)

            x = self.down_convs[i](x, edge_index, edge_weight)
            x = self.act(x)
            

            if i < self.depth: 
                xs += [x]
                edge_indices += [edge_index]
                edge_weights += [edge_weight]
            perms += [perm]
            
        mu2=self.muLay(x)
        sig2=self.sigLay(x)
        
        return mu2,sig2,xs,edge_indices,edge_weights,perms
        
        
    def reparametrize(self, mu, logvar):
      if self.training:
          return mu + torch.randn_like(logvar) * torch.exp(logvar)
      else:
          return mu  

        
    def decode(self,z,xs,edge_indices,edge_weights,perms):
        
        z=self.dec(z)
        for i in range(self.depth):
            j = self.depth - 1 - i

            res = xs[j]
          
            edge_index = edge_indices[j]
            edge_weight = edge_weights[j]
            perm = perms[j]


            up = torch.zeros_like(res)
            up[perm] = z
            z = res + up if self.sum_res else torch.cat((res, up), dim=-1)
            z = self.up_convs[i](z, edge_index, edge_weight)
            z = self.act(z) if i < self.depth - 1 else z
        del xs 
        torch.cuda.empty_cache()
        del edge_indices
        torch.cuda.empty_cache()
        del edge_weights
        torch.cuda.empty_cache()
        del perms
        torch.cuda.empty_cache()
        return z


    def augment_adj(self, edge_index, edge_weight, num_nodes):
        
        edge_index, edge_weight = add_self_loops(edge_index, edge_weight,
                                                 num_nodes=num_nodes)
        
        edge_index, edge_weight = sort_edge_index(edge_index, edge_weight,num_nodes)
        
        edge_index, edge_weight = spspmm(edge_index, edge_weight, edge_index,
                                         edge_weight, num_nodes, num_nodes,num_nodes)
        
        edge_index, edge_weight = remove_self_loops(edge_index, edge_weight)
        return edge_index, edge_weight      
    
    

    def forward(self,x,adj,lengs):
        
        mu2,sig2,xs,edge_indices,edge_weights,perms= self.encode(x,adj,lengs)  
        z = self.reparametrize(mu2,sig2) ## z = mu + eps*sigma 
        z2=self.decode(z,xs,edge_indices,edge_weights,perms)
        torch.cuda.empty_cache() 
        return z2, mu2, sig2  
 ``
  
            
                
                
                
                    `

**Comment by Anneqw** [comment]
Hi @ahariri13  I'm trying to reconstruct node features using Graph Neural Networks, which is pretty similar to your work, but I had a bad result. Currently I am using MSE loss in training, I would like to ask if this loss function is suitable. And could you please give me some advice about choosing a suiltable loss function？

Thanks a lot in advance ! :)
